internal/storage/metastorage/firestore/block_storage.go

```diff
@@ -0,0 +1,324 @@
+		}
+	}
+	return b.instrumentGetBlocksByHeights.Instrument(ctx, func(ctx context.Context) ([]*chainstorage.BlockMetadata, error) {
+		docRefs := make([]*firestore.DocumentRef, len(heights))
+		for i, height := range heights {
+			docRefs[i] = b.getCanonicalBlockDocRef(tag, height)
+		}
+		docs, err := b.client.GetAll(ctx, docRefs)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to get blocks by heights: %w", err)
+		}
+		blocks := make([]*chainstorage.BlockMetadata, len(heights))
+		expecting := 0
+		for i, doc := range docs {
+			if !doc.Exists() {
+				return nil, xerrors.Errorf(
+					"block metadata with height %d not found: %w",
+					heights[i], errors.ErrItemNotFound)
+			}
+			block, err := b.intoBlockMetadata(doc)
+			if err != nil {
+				return nil, xerrors.Errorf("failed to parse document snapshot into block metadata: %w", err)
+			}
+			if block.Height != heights[expecting] {
+				return nil, xerrors.Errorf(
+					"block metadata with height %d not found: %w",
+					heights[expecting], errors.ErrItemNotFound)
+			}
+			expecting = expecting + 1
+			blocks[i] = block
+		}
+		if expecting < len(heights) {
+			return nil, xerrors.Errorf(
+				"block metadata with height %d not found: %w",
+				heights[expecting], errors.ErrItemNotFound)
+		}
+		return blocks, nil
+	})
+}
+
+// GetLatestBlock implements internal.BlockStorage.
+func (b *blockStorageImpl) GetLatestBlock(ctx context.Context, tag uint32) (*chainstorage.BlockMetadata, error) {
+	return b.instrumentGetLatestBlock.Instrument(ctx, func(ctx context.Context) (*chainstorage.BlockMetadata, error) {
+		return b.getBlock(ctx, b.getLatestBlockDocRef(tag))
+	})
+}
+
+// PersistBlockMetas implements internal.BlockStorage.
+func (b *blockStorageImpl) PersistBlockMetas(ctx context.Context, updateWatermark bool, blocks []*chainstorage.BlockMetadata, lastBlock *chainstorage.BlockMetadata) error {
+	if len(blocks) == 0 {
+		return nil
+	}
+	return b.instrumentPersistBlockMetas.Instrument(ctx, func(sg_ctx context.Context) error {
+		sort.Slice(blocks, func(i, j int) bool {
+			return blocks[i].Height < blocks[j].Height
+		})
+		if err := parser.ValidateChain(blocks, lastBlock); err != nil {
+			return xerrors.Errorf("failed to validate chain: %w", err)
+		}
+
+		// Blocks with hash as key can be written in random order
+		bulkWriter := b.client.BulkWriter(sg_ctx)
+		for _, block := range blocks {
+			docRef := b.getBlockDocRef(block.Tag, block.Height, block.Hash)
+			_, err := bulkWriter.Set(docRef, b.fromBlockMetadata(block))
+			if err != nil {
+				return xerrors.Errorf("failed to add block %+v to BulkWriter: %w", block, err)
+			}
+		}
+		bulkWriter.End()
+
+		// Canonical blocks must be written in order to avoid canonical chain inconsistency with best effort
+		writeCount := len(blocks)
+		if updateWatermark {
+			// Append watermark block to be the last one to update
+			writeCount++
+		}
+		for c := 0; c < (writeCount+maxBulkWriteSize-1)/maxBulkWriteSize; c++ {
```

Comment:
nit: something like this would be a little bit more readable
```
firstChunkSize := maxBulkWriteSize
// First chunk as the residual
if writeCount > maxBulkWriteSize {
	firstChunkSize = writeCount % maxBulkWriteSize
}

for start, end := 0, firstChunkSize; start < writeCount; start, end = end, end+maxBulkWriteSize {
	if end > writeCount {
		end = writeCount
	}
	...
}```

Reply:
Great suggestion! Amended.